### PR TITLE
Refactor tenant

### PR DESCRIPTION
**Pull request by mpolden**
* Split tenant types into separate classes
* Stop using TenantId internally in controller-server
* Test variations of invalid tenant names and normalization

FYI @bjorncs

### CODE REVIEW

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/tenant/Tenant.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import com.yahoo.vespa.athenz.api.AthenzDomain;
+
+import java.util.Objects;
+
+/**
+ * A tenant in hosted Vespa.
+ *
+ * @author mpolden
+ */
+public abstract class Tenant {
```
Comment (bjorncs) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/tenant/Tenant.java]: I don't like the `.type()`, `.isAthenz()`, `.isUser()`, `in(AthenzDomain)` and `isUser(String)` members. The type is already modelled through inheritance, so I don't see any strong arguments for having an explicit `Type` enum. I prefer having these methods replaced with use of `instanceof` + casting.

**Review by bjorncs — COMMENTED** [review]
See comment

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/tenant/Tenant.java (hunk 1 above)
Comment (mpolden) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/tenant/Tenant.java]: Agreed, I'll replace it with explicit casts.

Comment (mpolden) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/tenant/Tenant.java]: PTAL

**Review by bjorncs — APPROVED** [review]

**Comment by jonmv** [comment]
Looks good :)

**Pull request merged**